TYP: Annotations in pandas/core/nanops.py

ShaharNaveh opened the pull request:
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Commit:
TYP: Annotations in pandas/core/nanops.py

jreback commented on pandas/core/nanops.py:
```diff
@@ -163,7 +166,9 @@ def _has_infs(result) -> bool:
 
-def _get_fill_value(dtype, fill_value=None, fill_value_typ=None):
+def _get_fill_value(
+    dtype: Dtype, fill_value: Any = None, fill_value_typ: Optional[str] = None
```
you can be more restrictive on the fill_value, I think this can only be Optional[Scalar], we almost *always* want to avoid Any

jreback commented on pandas/core/nanops.py:
```diff
@@ -1240,7 +1305,7 @@ def nancorr(a, b, method="pearson", min_periods=None):
+def get_corr_func(method: str):
     if method in ["kendall", "spearman"]:
         from scipy.stats import kendalltau, spearmanr
     elif callable(method):
```
can you add another conditions which checks that method in ['kendall', 'spearman', 'kendall'] *only*

MomIsBestFriend replied:
> can you add another conditions which checks that method in ['kendall', 'spearman', 'kendall'] _only_

Just to be sure, did you mean:
```python
["kendall", "spearman", "pearson"]
```
?

jreback replied:
yes

jreback commented on pandas/core/nanops.py:
```diff
@@ -1218,7 +1278,12 @@ def _zero_out_fperr(arg):
+def nancorr(
+    a: np.ndarray,
+    b: np.ndarray,
+    method: str = "pearson",
```
this needs to be Union[str, Callable[...]]

MomIsBestFriend replied:
mypy raises errors

jreback replied:
then something else is wrong

jreback commented on pandas/core/nanops.py:
```diff
@@ -1240,7 +1305,7 @@ def nancorr(a, b, method="pearson", min_periods=None):
 
 
-def get_corr_func(method):
+def get_corr_func(method: str):
```
this returns a Callable

jbrockmendel commented:
In the CI logs:
```
>           from scipy.stats import kendalltau, spearmanr
E           ModuleNotFoundError: No module named 'scipy'
```

For this to work you need to first do `from scipy import stats`, then do the specific imports.  I think a recent change in the test imports may be the reason why this PR was affected, but the import here definitely needs to be updated.

jbrockmendel commented:
apparently a test in test_nanops is missing a td.skip_if_no_scipy.  have you rebased recently?

MomIsBestFriend commented:
> have you rebased recently?

Yes, I have

jbrockmendel commented on pandas/core/nanops.py:
```diff
@@ -629,7 +654,7 @@ def _get_counts_nanvar(
     axis: Optional[int],
     ddof: int,
-    dtype=float,
+    dtype: Dtype = float,
```
@simonjayhawkins my read of Dtype is that `float` shouldn't be included.  Is this not-failing because np.dtype resolves to Any?

simonjayhawkins replied:
yes, definition of Dtype alias is being updated in #29046 to allow this.

MomIsBestFriend replied:
So keep the ```Dtype``` typing here? or remove it?

simonjayhawkins replied:
OK to keep IMO

jbrockmendel commented on pandas/core/nanops.py:
```diff
@@ -1075,7 +1131,13 @@ def nankurt(values, axis=None, skipna=True, mask=None):
+    skipna: bool = True,
+    min_count: int = 0,
+    mask: Optional[np.ndarray] = None,
+) -> Dtype:
```
Dtype as the return type looks sketchy.  are you sure about this?

jbrockmendel commented on pandas/core/nanops.py:
```diff
@@ -1218,7 +1280,9 @@ def _zero_out_fperr(arg):
 @disallow("M8", "m8")
-def nancorr(a, b, method="pearson", min_periods=None):
+def nancorr(
+    a: np.ndarray, b: np.ndarray, method="pearson", min_periods: Optional[int] = None,
```
can method be typed as str?

jbrockmendel commented on pandas/core/nanops.py:
```diff
@@ -1240,8 +1304,9 @@ def nancorr(a, b, method="pearson", min_periods=None):
-    if method in ["kendall", "spearman"]:
+def get_corr_func(method) -> Callable:
+    if method in ["kendall", "spearman", "pearson"]:
+        from scipy import stats
```
i think the CI issue is that this should be `import scipy.stats`

Commit:
Fixes for jbrockmendel review

Commit:
Added scipy to dep list of azure-macOS and azure-36-32bit

simonjayhawkins commented on pandas/core/nanops.py:
```diff
@@ -1088,18 +1150,14 @@ def nanprod(values, axis=None, skipna=True, min_count=0, mask=None):
     Returns
     -------
-    result : dtype
+    The product of all elements on a given axis. ( NaNs are treated as 1)
```
should be type and then description. same as  _maybe_null_out?

simonjayhawkins reviewed (commented):
comment

Commit:
Fixes for review

Commit:
Added dtype annotation

Commit:
Removed changes that are not typing related

MomIsBestFriend commented on pandas/core/nanops.py:
```diff
@@ -1240,8 +1310,8 @@ def nancorr(a, b, method="pearson", min_periods=None):
-def get_corr_func(method):
-    if method in ["kendall", "spearman"]:
+def get_corr_func(method) -> Callable:
+    if method in ["kendall", "spearman", "pearson"]:
```
This is causing too much trouble, I will open a separate PR dealing with it.

Commit:
Merge remote-tracking branch 'upstream/master' into TYP-core-nanops

jreback commented:
looks ok, @MomIsBestFriend there is the return type of nanprod that is incorrect (and not erroring) ?

MomIsBestFriend commented:
On my PC the type is

```python
<class 'numpy.float64'>
```

I think that's is because my CPU is 64-bit architecture, should I annotate it as
```python
Union[np.float32, np.float64]
```
to add support for 32-bit architecture CPUs as well?
or just to annotate it as ```float``` or ```np.float```?


cc @jreback @simonjayhawkins

simonjayhawkins commented:
> or just to annotate it as `float` or `np.float`?

just `float` for now, since np.float resolves to `Any` without stubs for numpy.

Commit:
Merge remote-tracking branch 'upstream/master' into TYP-core-nanops

Commit:
Removed 'Dtype' as a return type

Commit:
Merge remote-tracking branch 'upstream/master' into TYP-core-nanops

Commit:
Merge remote-tracking branch 'upstream/master' into TYP-core-nanops

Commit:
Merge remote-tracking branch 'upstream/master' into TYP-core-nanops

Commit:
Merge remote-tracking branch 'upstream/master' into TYP-core-nanops

Commit:
Reverted typing

Commit:
List issues

Commit:
Merge remote-tracking branch 'upstream/master' into TYP-core-nanops

jreback commented on pandas/core/nanops.py:
```diff
@@ -129,7 +132,7 @@ def f(values, axis=None, skipna=True, **kwds):
 
 
-def _bn_ok_dtype(dt, name: str) -> bool:
+def _bn_ok_dtype(dt: Dtype, name: str) -> bool:
```
can you dt -> dtype

MomIsBestFriend replied:
Addressed in https://github.com/pandas-dev/pandas/pull/30461/commits/328356467eb6d4b636a2e5acf696b059648ce5b5

jreback commented on pandas/core/nanops.py:
```diff
@@ -1218,7 +1283,9 @@ def _zero_out_fperr(arg):
 @disallow("M8", "m8")
-def nancorr(a, b, method="pearson", min_periods=None):
+def nancorr(
+    a: np.ndarray, b: np.ndarray, method="pearson", min_periods: Optional[int] = None,
```
method: str ?

jreback replied:
or Union[str, Callable]

MomIsBestFriend replied:
In https://github.com/pandas-dev/pandas/pull/30461/commits/7442310a426f9160abab5e201c10b8ad4590b25d I removed all annotations of ```str``` and ```Callable``` that were added in this PR.

No matter what annotation I try, mypy will complain, I will maybe be able to address this annotation in another PR, but I think that is out of scope for this PR at least.

jreback commented on pandas/core/nanops.py:
```diff
@@ -1240,7 +1307,7 @@ def nancorr(a, b, method="pearson", min_periods=None):
 
 
-def get_corr_func(method):
+def get_corr_func(method) -> Callable:
```
method: Union[str, Callable]

MomIsBestFriend replied:
In https://github.com/pandas-dev/pandas/pull/30461/commits/7442310a426f9160abab5e201c10b8ad4590b25d I removed all annotations of ```str``` and ```Callable``` that were added in this PR, it just gave me too much trouble.

jreback reviewed (changes requested):
pls rebase as well

Commit:
Merge remote-tracking branch 'upstream/master' into TYP-core-nanops

Commit:
Changed the name of "dt" to "dtype"

Ref: https://github.com/pandas-dev/pandas/pull/30461#discussion_r368243516

Commit:
Removed any newly added annotation of Callable/str

This is because no matter what typing annotation combination of "method"
is being used "Union[Callable, str]" or "Callable" or "str" mypy is
raising errors

MomIsBestFriend commented:
> pls rebase as well

Rebased.

Commit:
Merge remote-tracking branch 'upstream/master' into TYP-core-nanops

jreback commented:
lgtm. @simonjayhawkins or @WillAyd if any comments. let's merge if good rather than add things to this one. ideally we keep basic typing PRs non-controversial and separate out the controversial to other PRs.

WillAyd commented:
Thanks @MomIsBestFriend
